Landmark parser

vesperlou opened the pull request:
# Issue

Implement landmark parser to parse landmarks in pbf and store them in database.

## Requirements / Relations

[Landmark Routing 2: Parse Landmarks](https://github.com/valhalla/valhalla/issues/4133)

vesperlou commented:
Hi @kevinkreiser @nilsnolde , I just finished task 2 for parsing and storing landmarks. Tests all passed :)
Not sure if I should split these stuff to seperate files like `admin builder`. Now the database, parser, and general function are all in `landmark_database_builder.cc` and `.h`

vesperlou commented:
Another question is in my tests the landmark `type` are parsed directly to `string` not `uint_64`. Is real data different or did I miss some steps to convert `string` to `uint_64`?

kevinkreiser commented on valhalla/mjolnir/landmark_database_builder.h:
```diff
@@ -66,6 +65,14 @@ struct LandmarkDatabase {
   void release_prepared_stmt();
 };
 
+// class?
+struct LandmarkParser {
+public:
+  static std::vector<Landmark> Parse(const std::vector<std::string>& input_files);
+};
```
you can move this to the implementation file, its not needed for the public api, only the build function is needed which will make use of this parser callback thing, similar to the admin processing

kevinkreiser replied:
oh i see this is the class analog. so in this case because you dont need any state, you can totally just remove this and put the guts of the work into the build function! sorry for the misunderstanding

vesperlou replied:
done.

vesperlou commented on src/mjolnir/landmark_database_builder.cc:
```diff
@@ -169,10 +200,16 @@ std::vector<Landmark> LandmarkDatabase::get_landmarks_in_bounding_box(const doub
-    const char* type = reinterpret_cast<const char*>(sqlite3_column_text(bounding_box_stmt, 1));
+
+    LandmarkType type = LandmarkType::NA;
+    if (sqlite3_column_type(bounding_box_stmt, 1) != SQLITE_NULL) {
+      type = static_cast<LandmarkType>(sqlite3_column_int(bounding_box_stmt, 1));
+    }
+
```
I updated landmark type to enum of uint8_t which introduces some data type conversions like here.
I wonder if this could be a problem.

vesperlou commented on src/mjolnir/landmark_database_builder.cc:
```diff
@@ -107,8 +138,8 @@ bool LandmarkDatabase::insert_landmark(const Landmark& landmark) {
 
-  if (landmark.type != "") {
-    sqlite3_bind_text(insert_stmt, 2, landmark.type.c_str(), landmark.type.length(), SQLITE_STATIC);
+  if (landmark.type != LandmarkType::NA) {
+    sqlite3_bind_int(insert_stmt, 2, static_cast<int>(landmark.type));
```
and here too

kevinkreiser commented on src/mjolnir/landmark_database_builder.cc:
```diff
@@ -1,8 +1,39 @@
+const std::string landmark_db = "landmarks.db";
+
+static const std::unordered_map<std::string, LandmarkType> str_type_map = {
+    {"parking", LandmarkType::parking},
+    {"bench", LandmarkType::bench},
+    {"parking_space", LandmarkType::parking_space},
```
i dont think any of these make sense really. maybe the first one whic could be a parking lot, but in reality it could but just some spaces on the side of the road. remember we are trying to call out things that are important, things people wont have to think about to see to navigate by

kevinkreiser replied:
oh and i should say by any i mean the top 3 here. if you noticed github lets you mark multiple lines but it doesnt actually highlight them when viewing the commment, but see above how it says +11 to +13. im not sure if you were aware of this github gotcha :smile:

vesperlou replied:
yes I can see the highlighted 3 lines. 
For the landmrks here I just chose the 15 most popular ones. it make sense to carefully choose some obvious and straightfoward things. but from the amenity website most of the obvious landmarks are not very popular. we dont consider that as an issue right now?

vesperlou replied:
correct: some of the obvious landmarks are still popular like "restaurant". so popularity is not an issue :)

kevinkreiser commented on src/mjolnir/landmark_database_builder.cc:
```diff
@@ -1,8 +1,39 @@
+    {"parking_space", LandmarkType::parking_space},
+    {"place_of_worship", LandmarkType::place_of_worship},
+    {"restaurant", LandmarkType::restaurant},
+    {"waste_basket", LandmarkType::waste_basket},
```
yeah no way are we using a trash can to navigate :smile:

vesperlou replied:
make sense haha. trash cans can move, not good either.

kevinkreiser commented on src/mjolnir/landmark_database_builder.cc:
```diff
@@ -1,8 +1,39 @@
+    {"place_of_worship", LandmarkType::place_of_worship},
+    {"restaurant", LandmarkType::restaurant},
+    {"waste_basket", LandmarkType::waste_basket},
+    {"bicycle_parking", LandmarkType::bicycle_parking},
```
same for this, doesnt seem relevant. maybe you are thinking about pedestrian routing? i suppose its possible that on unnamed trails smaller things like these could be worth using for navigation points?

kevinkreiser commented on src/mjolnir/landmark_database_builder.cc:
```diff
@@ -1,8 +1,39 @@
+    {"fast_food", LandmarkType::fast_food},
+    {"cafe", LandmarkType::cafe},
+    {"fuel", LandmarkType::fuel},
+    {"shelter", LandmarkType::shelter},
+    {"recycling", LandmarkType::recycling},
+    {"toilets", LandmarkType::toilets},
```
these ones dont make much sense either

kevinkreiser commented:
> Now the database, parser, and general function are all in landmark_database_builder.cc and .h

in my opinion this is good. id prefer to keep all the stuff in one place. we already have a huge project the more like things that can go in one source the better imho


> Or I should encapsulate enum with uint8_t like enum class LandmarkType : uint8_t { parking = 0, restaurant = 1, };
yes thats what you should do. to do that we have a bunch of similar conversions using unordered maps. here is an example: https://github.com/valhalla/valhalla/blob/c3089cb52a05a61f3312747f88c49453b5069937/valhalla/baldr/graphconstants.h#L135-L161
you want the conversion in both directions

kevinkreiser commented:
i think for now we can focus on landmark navigation for motor vehicle routing and just ingest the larger most obvious types of amenities, bars, pubs, libraries, schools, restaurants, gas stations, banks, mechanics shops, firestation, post office, hotel. basically big buildings that shoudl be obviously marked so that the person doesnt have to look for them to navigate. think of it as something that should be more obvious than the road name sign itself

kevinkreiser commented on src/mjolnir/landmark_database_builder.cc:
```diff
@@ -220,6 +257,90 @@ void LandmarkDatabase::release_prepared_stmt() {
+  }
+
+  // store landmarks in database
+  LandmarkDatabase db(landmark_db, false);
+
+  for (const auto& landmark : landmarks) {
+    if (!db.insert_landmark(landmark)) {
```
do this inside the callback for each one you parse. there is no sense in allocating a giant vector when you can just push a landmark into the db immediately as soon as its parsed. this means the parser needs the landmarkdatabase as a member though of course

vesperlou replied:
good idea. I separated parser and insert because I wanted to test if parser is correct. now I know it is so can put insert into callback directly.

vesperlou commented:
> > Now the database, parser, and general function are all in landmark_database_builder.cc and .h
> 
> in my opinion this is good. id prefer to keep all the stuff in one place. we already have a huge project the more like things that can go in one source the better imho
> 
> > Or I should encapsulate enum with uint8_t like enum class LandmarkType : uint8_t { parking = 0, restaurant = 1, };
> > yes thats what you should do. to do that we have a bunch of similar conversions using unordered maps. here is an example: https://github.com/valhalla/valhalla/blob/c3089cb52a05a61f3312747f88c49453b5069937/valhalla/baldr/graphconstants.h#L135-L161
> > 
> > you want the conversion in both directions

Thanks for the reference! I already implemented string_to_landmark in the implementation file. should I implement the two func in the header like this one? Or implementation file is fine for they are not publicly exposed from my view.

vesperlou commented:
> i think for now we can focus on landmark navigation for motor vehicle routing and just ingest the larger most obvious types of amenities, bars, pubs, libraries, schools, restaurants, gas stations, banks, mechanics shops, firestation, post office, hotel. basically big buildings that shoudl be obviously marked so that the person doesnt have to look for them to navigate. think of it as something that should be more obvious than the road name sign itself

Make sense. maybe later we can add things for pedastrian navigation if everything works well.

kevinkreiser commented:
> should I implement the two func in the header like this one
for now its ok in the source but we'll need the other direction if we want to use the string in our narrative code "turn left at the church" where "church" comes from your enum to string conversion

vesperlou commented:
> > should I implement the two func in the header like this one
> 
> for now its ok in the source but we'll need the other direction if we want to use the string in our narrative code "turn left at the church" where "church" comes from your enum to string conversion

No extra efforts to put it in the header then! done.

vesperlou commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -220,6 +228,80 @@ void LandmarkDatabase::release_prepared_stmt() {
+  }
+  virtual ~landmark_callback() {
+  }
+
+  virtual void
+  node_callback(const uint64_t /*osmid*/, double lng, double lat, const OSMPBF::Tags& tags) override {
+    Landmark landmark;
+
+    for (const auto& tag : tags) {
+      if (tag.first == "amenity") {
+        // if amenity is empty will return LandmarkType::null
+        landmark.type = string_to_landmark_type(tag.second);
+      }
+      if (tag.first == "name" && !tag.second.empty()) {
+        landmark.name = tag.second;
+      }
+      landmark.lng = lng;
+      landmark.lat = lat;
+    }
+    // insert parsed landmark directly into database
+    db_.insert_landmark(landmark);
+  }
+
+  virtual void changeset_callback(const uint64_t changeset_id) override {
+  }
+
+  virtual void way_callback(const uint64_t /*osmid*/,
+                            const OSMPBF::Tags& /*tags*/,
+                            const std::vector<uint64_t>& /*nodes*/) override {
+    LOG_WARN("way callback shouldn't be called!");
+  }
+
+  virtual void relation_callback(const uint64_t /*osmid*/,
+                                 const OSMPBF::Tags& /*tags*/,
+                                 const std::vector<OSMPBF::Member>& /*members*/) override {
+    LOG_WARN("relation callback shouldn't be called!");
+  }
+
+  LandmarkDatabase& db_;
+};
```
In admin builder the callback struct is put in anonymous namespace. should I do the same here? don't know the benefits though

nilsnolde replied:
Jep we should do the same here. In this case there'd be hardly a benefit, other than making it clear that it's not part of the public interface and rather implementation. If it's inside our named namespaces, I'd expect the callback struct's definition to be inside the header file too so I could import it into other files. Since we only use that struct in this .cc file, we should put it into a anonymous namespace.

kevinkreiser replied:
unnamed namespace is best for two reasons:

1. this is not in the valhalla/mjolnir namespace because we want it to remain an implementation detail of the library and not something in the public interface
2. adding it to a namespace makes it local to that translation unit and as such avoids potential naming clashes when linking

vesperlou commented:
@kevinkreiser @nilsnolde I checked the project roadmap and found there is still one thing to do: decide whether parsing landmarks should be in a separate executable and be run separately. WDYT?

nilsnolde commented on test/gurka/test_landmarks.cc:
```diff
@@ -13,15 +13,46 @@ using namespace valhalla::mjolnir;
+  const gurka::nodes nodes = {
+      {"A", {{"name", ""}, {"amenity", "university"}}},
+      {"B", {{"name", "hai di lao"}, {"amenity", "restaurant"}}},
+      {"C", {{"name", "ke ji lu"}, {"amenity", ""}}},
```
IMO this should be a degenerate case we want to throw early, as it has no meaning without a type.

nilsnolde replied:
let's also include a node that shouldn't be saved. Maybe we can make this more realistic by e.g.

```
  const std::string ascii_map = R"(
      a-------b------c
          A    B     C
    )";
```
So A, B, C would be on the side of the road, and a, b, c are simply graph nodes without any tags (which shouldn't appear in the db, but I think do currently, see above).

vesperlou replied:
good example I will update that.

nilsnolde commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -220,6 +228,80 @@ void LandmarkDatabase::release_prepared_stmt() {
+  node_callback(const uint64_t /*osmid*/, double lng, double lat, const OSMPBF::Tags& tags) override {
+    Landmark landmark;
+
+    for (const auto& tag : tags) {
+      if (tag.first == "amenity") {
+        // if amenity is empty will return LandmarkType::null
+        landmark.type = string_to_landmark_type(tag.second);
+      }
+      if (tag.first == "name" && !tag.second.empty()) {
+        landmark.name = tag.second;
+      }
+      landmark.lng = lng;
+      landmark.lat = lat;
+    }
+    // insert parsed landmark directly into database
+    db_.insert_landmark(landmark);
```
doesn't this save _all_ nodes right now?! 

Also I'd not handle the case where `amenity` has no value. I'm not even sure that can happen in actual OSM data (but surely can in custom OSM data). Even there is smth like that, it has no value for us IMO.

nilsnolde replied:
We could instead do a LOG_ERROR when there's an amenity node without a value.

kevinkreiser replied:
agreed, it has to have a type otherwise its meaningless

vesperlou replied:
thanks for pointing this out! I totally forgot to include landmark nodes only and made a premise that not all landmark nodes have amenity value which is incorrect. will fix that.

vesperlou replied:
a further point: in the landamark database should we only store nodes that belong to `enum class LandmarkType`? Now the list contians 21 popular landmark types and others will be translated into `LandmarkType::null` which is useless in navigation. plus we can shorten execution time and save storage by storing "known" nodes only.

kevinkreiser replied:
exactly right, we should skip anything that is useless, no point in even looking at it if we cant parse the string into one of the enums that we accept

vesperlou replied:
got it. done.

nilsnolde commented on valhalla/mjolnir/landmark_builder.h:
```diff
@@ -0,0 +1,164 @@
+}
+
+struct Landmark {
+  std::string name;
+  LandmarkType type;
```
the way we're currently using this struct, we should default at least the name & type or you'll get random stuff in there if they're never set.

vesperlou replied:
ah right I forgot default / initialization again. let's set default name as `"null"` and default type as `LandmarkType::null` :D

kevinkreiser replied:
i disagree here. i think we shouldnt even h ave this struct but rather for the insert we can take the parameters in the function (why make a copy when we dont need to) and then when returning we can just return a tuple. as far as defaults, name can either exist or be empty we dont need a name for all amenities the type is good enough for the majority. default location makes no sense, we should never have a landmark that doesnt have a location. again i think we can avoid this by simply removing this structure completely from the interface

vesperlou replied:
I can remove the struct. it simply makes the code prettier and easier so no problem. As for default value of name I think having a default `"null" `or `""` or none at all are interchangeable to me. which one is better from your opinion?
but I don't know how to check validity of lat and lng. other implementation files also don't deal with that so I assume it is either checked in parser or just doesn't exist, or both. so for `lat` and `lng` maybe we can keep the current with no defaults. WDYT? @nilsnolde

kevinkreiser replied:
in the database, it can be NULL but in code it has to be the empty string. therefor to make our lives easier it could also be empty string in the database

kevinkreiser commented:
> I checked the project roadmap and found there is still one thing to do: decide whether parsing landmarks should be in a separate executable and be run separately. WDYT?

originally i thought it should be just like admin building that we run it once in a separate process but now im thinking it might be nice to just turn on for regular tile builds. i was thinking like this


```json
{
  "mjolnir": {
    ...
    "parse_landmarks": false, // if you turn this on the regular osm parsing callback parses the nodes for landmarks
    "landmarks_db": the/path/to/the/db.sqlite, // this would be the location it writes to if parse is true and it will be the location it reads from in the enhancer
    ...
  }
}
```

this would mean a couple changes for this pr, but essentially we'd just add a tiny bit of code to the existing pbf graphparser.

alternatively we could make another executable called `valhalla_build_landmarks` and have it run separately. i will say that the reason we do that for admins is because they dont change that often and actually we dont like to take changes from them too often because people will break them on occasion. with landmarks i think we'd always like to get the latest that exists in the data regardless which makes me think it should just happen (still optionally) in the parse stage with the rest of the graph data

@nilsnolde what do you think about that?

nilsnolde commented:
Hm I think our bigger idea was that eventually all these graph enhancements like landmarks/transit/elevation would be best to have as external executables rather than cramming it all into the regular build pipeline. Comes with an overhead (parsing PBF again in this case), but that’d be fine IMO. 

So I think I’d advocate to keep it separate and have it enrich an existing built graph if possible. But I might overlook smth crucial?

nilsnolde commented:
We could still have it in build pipeline as very last steps so people don’t get too confused on what they need to run in which order, but IMO rather not tightly coupled with the initial graph build.

kevinkreiser commented:
separate is totally fine but yeah if I were running the system I'd hate to spend a couple hours parsing data before being able to run the tile build. my suggestion is you can still make and bring your sqlite file from. wherever outside process but we could offer the option to build it along with the regular parsing step so that it's up to date and doest require extra time to make. allow both use cases in other words

vesperlou commented:
ok I will allow both use cases. it makes sense to offer choices for various situations and is what I would do. 
just need to fidure out how. already have some clues :D

nilsnolde commented on test/gurka/test_landmarks.cc:
```diff
@@ -13,16 +13,58 @@ using namespace valhalla::mjolnir;
+    )";
+
+  const gurka::nodes nodes = {
+      {"A", {{"landmark", "yes"}, {"name", ""}, {"amenity", "university"}}},
+      {"B", {{"landmark", "yes"}, {"name", "hai di lao"}, {"amenity", "restaurant"}}},
+      {"C", {{"landmark", "yes"}, {"name", "ke ji lu"}, {"amenity", ""}}}, // no amenity, log error
+      {"D", {{"landmark", "yes"}, {"name", "wan da"}, {"amenity", "cinema"}}},
```
also  don't forget to remove the `landmark` in the tests

vesperlou replied:
meeting ends in one hour. we can discuss here @kevinkreiser @nilsnolde 
I think the `landmark` key is used to select landmark nodes as you suggested earlier we should only parse landmarks nodes otherwise there are too many. without landmark key how can we achieve that?

nilsnolde replied:
Only the nodes which have a "amenity" tag having a value of the map you specified in your new header (hospital, restaurant etc) are the nodes we want to save. earlier you were just saving _every_ node. however, the `landmark` tag is useless for us. the gist is, we want to define what we consider a "landmark" and not leave that definition to random OSM mappers. does that make sense now?

vesperlou replied:
ah right I understand now. thanks!

nilsnolde commented on test/gurka/test_landmarks.cc:
```diff
@@ -13,16 +13,58 @@ using namespace valhalla::mjolnir;
+  };
+
+  const gurka::ways ways = {
+      {"ab", {}},
+      {"bc", {}},
+      {"cd", {}},
```
let's give it at least a `{"highway", "residential"}` or some other highway attribute

vesperlou replied:
done

vesperlou commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -1,10 +1,67 @@
+  virtual void
+  node_callback(const uint64_t osmid, double lng, double lat, const OSMPBF::Tags& tags) override {
+    // parse "landmark" nodes only
+    auto iter = tags.find("landmark");
+    if (iter != tags.cend() && iter->second == "yes") {
```
no problem very clear to me :D

vesperlou commented:
@nilsnolde @kevinkreiser I finished addressing all issues discussed today, feel free to check :D
seems like CI is down for a while, clicking details merely returns 421 Error. maybe we should rerun the failed CI task.

I searched tests for command line but didn't find any reference, so I did a manual test to proove it works as expected. should I write test for `valhalla_build_landmarks` cli? 
and I think the current tests look a bit messy. the first test is about building database and bounding box only so I don't want to mix its data with others (mixing could be a problem in design of bounding box query which is often used to check correctness). In fact, I don't want to mix data from any single test, but all tests should have database initialization and file removal. repeating code is ugly... is there a better way to write the tests?

nilsnolde commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -1,10 +1,63 @@
+
+  virtual void
+  node_callback(const uint64_t /*osmid*/, double lng, double lat, const OSMPBF::Tags& tags) override {
+    valhalla::mjolnir::Landmark landmark;
```
this wasn't addressed yet: we don't need this intermediate struct at all. just put the properties into the `insert_landmark` call.

nilsnolde replied:
let's keep it unresolved until we can see it's done :) best to let the reviewers resolve comments, then they don't lose the overview of their own comments

vesperlou replied:
no problem. thanks for the tip! 
and it's fixed now.

nilsnolde commented on src/mjolnir/valhalla_build_landmarks.cc:
```diff
@@ -0,0 +1,70 @@
+    valhalla::midgard::logging::Configure(logging_config);
+  }
+
+  // if (!valhalla::mjolnir::BuildAdminFromPBF(pt.get_child("mjolnir"), input_files)) {
```
remove the comment

vesperlou replied:
done

nilsnolde reviewed (changes requested):
To expand a bit on the executable testing: we don't do that, though it can be done, but it's fairly much effort and not much point as long as 99% of the logic is outside the executable, which is why we advocate for that.

There's a few more things to address.

nilsnolde commented:
FYI, the OSX failure will be fixed with #4209

nilsnolde commented on test/gurka/test_landmarks.cc:
```diff
@@ -2,60 +2,142 @@
+      {"B", {{"name", "hai di lao"}, {"amenity", "restaurant"}}},
+      {"C", {{"name", "ke ji lu"}, {"amenity", ""}}}, // no amenity, shouldn't be stored
+      {"D", {{"name", "wan da"}, {"amenity", "cinema"}}},
+      {"E", {{"name", "zhong lou"}, {"amenity", "monument"}}}, // not in list, shouldn't be stored
```
(accidentally?) you gave a VERY good example of OSM stuff we'd also like to save apart from `amenity`. See [`historic` key](https://taginfo.openstreetmap.org/keys/historic#values) which also contains `monument`. If the name is descriptive enough that's the perfect guidance POI:)

But let's shelve that for a later day: https://github.com/valhalla/valhalla/issues/4210

vesperlou replied:
you remind me it's not an accident :) I was discovering osm key-values and found `monument`, thought it could be interesting to have it as landmarks but forgot to discuss with you. no problem let's discuss later.

vesperlou commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -104,28 +157,21 @@ LandmarkDatabase::LandmarkDatabase(const std::string& db_name, bool read_only)
     : pimpl(new db_pimpl(db_name, read_only)) {
 }
 
-void LandmarkDatabase::insert_landmark(const Landmark& landmark) {
+void LandmarkDatabase::insert_landmark(const std::string& name,
+                                       const LandmarkType& type,
+                                       const double lng,
+                                       const double lat) {
   auto* insert_stmt = pimpl->insert_stmt;
   if (!insert_stmt)
     throw std::logic_error("Sqlite database connection is read-only");
 
   sqlite3_reset(insert_stmt);
   sqlite3_clear_bindings(insert_stmt);
 
-  if (landmark.name != "") {
-    sqlite3_bind_text(insert_stmt, 1, landmark.name.c_str(), landmark.name.length(), SQLITE_STATIC);
-  } else {
-    sqlite3_bind_null(insert_stmt, 1);
-  }
-
-  if (landmark.type != "") {
-    sqlite3_bind_text(insert_stmt, 2, landmark.type.c_str(), landmark.type.length(), SQLITE_STATIC);
-  } else {
-    sqlite3_bind_null(insert_stmt, 2);
-  }
-
-  sqlite3_bind_double(insert_stmt, 3, landmark.lng);
-  sqlite3_bind_double(insert_stmt, 4, landmark.lat);
+  sqlite3_bind_text(insert_stmt, 1, name.c_str(), name.length(), SQLITE_STATIC);
+  sqlite3_bind_int(insert_stmt, 2, static_cast<int>(type));
+  sqlite3_bind_double(insert_stmt, 3, lng);
+  sqlite3_bind_double(insert_stmt, 4, lat);
```
I removed the empty checks for name and type here because in common use cases `insert_landmark` should only be called by `node_callback` where checks and initializations already happen.

nilsnolde commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -153,10 +199,17 @@ std::vector<Landmark> LandmarkDatabase::get_landmarks_in_bounding_box(const doub
-    const char* type = reinterpret_cast<const char*>(sqlite3_column_text(bounding_box_stmt, 1));
+
+    LandmarkType type = LandmarkType::null;
+    if (sqlite3_column_type(bounding_box_stmt, 1) != SQLITE_NULL) {
+      type = static_cast<LandmarkType>(sqlite3_column_int(bounding_box_stmt, 1));
+    }
```
it shouldn't happen that the type is empty. IIRC we throw away nodes that don't have an `amenity` type no? I'd consider that a bug if that happens, but pretty sure it can't

kevinkreiser replied:
agree, its probably not needed to check this

nilsnolde commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -168,5 +221,44 @@ std::vector<Landmark> LandmarkDatabase::get_landmarks_in_bounding_box(const doub
+  if (!filesystem::exists(parent_dir) && !filesystem::create_directories(parent_dir)) {
+    LOG_ERROR("Can't create parent directory " + parent_dir.string());
+    return false;
+  }
```
we should remove the database right here if it exists

kevinkreiser replied:
yeah i was trying to say before that if we open the database in non-read-only mode and its there we should just blow it away. that can happen inside the database construction imho, unless we see some future need to be able to append to the database?

vesperlou replied:
agree. for now it makes sense to allow writing only when a database is created. I'll update the code.

nilsnolde reviewed (changes requested):
Sorry, now with the new changes I somehow have quite a few more questions.

kevinkreiser commented on valhalla/mjolnir/landmark_builder.h:
```diff
@@ -0,0 +1,129 @@
+
+// obvious landmark types for vehicle routing
+enum class LandmarkType : uint8_t {
+  null = 0,
```
```suggestion
```

kevinkreiser commented on valhalla/mjolnir/landmark_builder.h:
```diff
@@ -0,0 +1,129 @@
+       {"cinema", LandmarkType::cinema}};
+
+  auto it = string_to_landmark_type.find(s);
+  if (it != string_to_landmark_type.cend()) {
+    return it->second;
+  }
+  return LandmarkType::null;
```
we can just disallow conversion when its unknown
```suggestion
  if (it == string_to_landmark_type.cend())
    throw std::runtime_error("unknown landmark type");
  return it->second;
```

vesperlou replied:
I just realize this function is called in `node_callback` to check every node with amenity. There are many amenity values not included in the list so we shouldn't throw. WDYT? @kevinkreiser if we remove LandmarkType::null how to do the check?

vesperlou replied:
ah we can use try catch

kevinkreiser replied:
exactly

kevinkreiser commented on valhalla/mjolnir/landmark_builder.h:
```diff
@@ -0,0 +1,129 @@
+  return LandmarkType::null;
+}
+
+inline std::string landmark_type_to_string(LandmarkType t) {
+  static const std::unordered_map<uint8_t, std::string> landmark_type_to_string =
+      {{static_cast<uint8_t>(LandmarkType::null), "na"},
+       {static_cast<uint8_t>(LandmarkType::restaurant), "restaurant"},
+       {static_cast<uint8_t>(LandmarkType::school), "school"},
+       {static_cast<uint8_t>(LandmarkType::fast_food), "fast_food"},
+       {static_cast<uint8_t>(LandmarkType::cafe), "cafe"},
+       {static_cast<uint8_t>(LandmarkType::fuel), "fuel"},
+       {static_cast<uint8_t>(LandmarkType::bank), "bank"},
+       {static_cast<uint8_t>(LandmarkType::pharmacy), "pharmacy"},
+       {static_cast<uint8_t>(LandmarkType::kindergarten), "kindergarten"},
+       {static_cast<uint8_t>(LandmarkType::bar), "bar"},
+       {static_cast<uint8_t>(LandmarkType::hospital), "hospital"},
+       {static_cast<uint8_t>(LandmarkType::post_office), "post_office"},
+       {static_cast<uint8_t>(LandmarkType::pub), "pub"},
+       {static_cast<uint8_t>(LandmarkType::clinic), "clinic"},
+       {static_cast<uint8_t>(LandmarkType::police), "police"},
+       {static_cast<uint8_t>(LandmarkType::fire_station), "fire_station"},
+       {static_cast<uint8_t>(LandmarkType::car_wash), "car_wash"},
+       {static_cast<uint8_t>(LandmarkType::charging_station), "charging_station"},
+       {static_cast<uint8_t>(LandmarkType::college), "college"},
+       {static_cast<uint8_t>(LandmarkType::university), "university"},
+       {static_cast<uint8_t>(LandmarkType::theatre), "theatre"},
+       {static_cast<uint8_t>(LandmarkType::cinema), "cinema"}};
+
+  auto it = landmark_type_to_string.find(static_cast<uint8_t>(t));
+  if (it != landmark_type_to_string.cend()) {
+    return it->second;
+  }
+  return "null";
+}
```
i know i said we would need this eventually but we dont need it yet right? on second thought i think we will never need it because these things will need to be localized to be used in directions like "turn left at the gas station". we have a wholly other way of getting those types of things into the response so i think we can remove this unless there is some other use?
```suggestion
```

vesperlou replied:
you are right, we don't need it yet. will delete it. 
how do you plan to do the wholly other way?

kevinkreiser replied:
its a large topic, once we have this stuff in the tiles we can go over it. if you are interested now you can go look at https://github.com/valhalla/valhalla/blob/master/locales/README.md for a start

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -153,10 +199,17 @@ std::vector<Landmark> LandmarkDatabase::get_landmarks_in_bounding_box(const doub
     const char* name = reinterpret_cast<const char*>(sqlite3_column_text(bounding_box_stmt, 0));
-    const char* type = reinterpret_cast<const char*>(sqlite3_column_text(bounding_box_stmt, 1));
+
+    LandmarkType type = LandmarkType::null;
+    if (sqlite3_column_type(bounding_box_stmt, 1) != SQLITE_NULL) {
+      type = static_cast<LandmarkType>(sqlite3_column_int(bounding_box_stmt, 1));
+    }
```
```suggestion
    int landmark_type = -1;
    if (sqlite3_column_type(bounding_box_stmt, 1) != SQLITE_NULL) {
      landmark_type = sqlite3_column_int(bounding_box_stmt, 1);
    }
```

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -153,10 +199,17 @@ std::vector<Landmark> LandmarkDatabase::get_landmarks_in_bounding_box(const doub
     double lat = sqlite3_column_double(bounding_box_stmt, 3);
-    landmarks.emplace_back(Landmark{name, type, lng, lat});
+
+    landmarks.emplace_back(std::make_tuple(name, type, lng, lat));
```
```suggestion
    landmarks.emplace_back(std::make_tuple(name, static_cast<LandmarkType>(landmark_type), lng, lat));
```

vesperlou replied:
got it. thanks

kevinkreiser commented:
this looks just about done, but we should very much vet the list of amenities before we call it finished because those will be settled for ever once we bake them into tiles. maybe we should shorten the list to just the ones we are super sure about and add ones as we encounter useful examples.

anyway other than some nitpicks its there

kevinkreiser commented on valhalla/mjolnir/landmark_builder.h:
```diff
@@ -0,0 +1,129 @@
+  fire_station = 15,
+  car_wash = 16,
+  charging_station = 17,
+  college = 18,
+  university = 19,
```
these things are probably too large to use to navigate, id drop them

vesperlou replied:
make sense

kevinkreiser commented on valhalla/mjolnir/landmark_builder.h:
```diff
@@ -0,0 +1,90 @@
+namespace mjolnir {
+static const std::string default_landmark_name = "";
+
+// obvious landmark types for vehicle routing
+enum class LandmarkType : uint8_t {
+  restaurant = 1,
+  fast_food = 2,
+  cafe = 3,
+  fuel = 4,
+  bank = 5,
+  pharmacy = 6,
+  kindergarten = 7,
+  bar = 8,
+  hospital = 9,
+  post_office = 10,
+  pub = 11,
+  clinic = 12,
+  police = 13,
+  fire_station = 14,
+  car_wash = 15,
+  theatre = 16,
+  cinema = 17,
+  casino = 18,
+};
```
i disagree that some of these will be obvious, in fact i think some of them require a name. id really like to stress that reading what i wrote in the initial issue #4037 regarding landmarks, specifically the visibility/verbosity/specificity sections. we aren't making a poi database we are trying to pick features that might stand out more than a road sign because at the decision point in the route you have a high pressure situation where you must decide quickly. if we are at all unsure about a specific amenity type then we should err on not using it. anyway here are some of my thoughts on this inline:

```suggestion
// obvious landmark types for vehicle routing
enum class LandmarkType : uint8_t {
  // these will almost always be obvious by their function and don't require a name
  fuel = 1,
  post_office = 2,
  police = 3,
  fire_station = 4,
  car_wash = 5,

  // could be obvious if its a chain but could also just be a building
  restaurant = 6,
  // this will always be a chain in which case the name is a required distinguisher (as they are often next to each other on either side of a street)
  fast_food = 7,
  // this could be a chain making it easier to distinguish (starbucks) but could be a one off cafe in which case the name will be required
  cafe = 8,
  // on the fence about this one, these are often obvious without a name, maybe should go in the upper list?
  bank = 10,
  // same with this one, in some countries these are obvious and have the little plus sign, in america its not the case and they can be different chains or could be part of a grocery store or anything... speaking of grocery store, why isnt that on the list? also why arent department stores on the list, surely stores would be a good landmark to use (again when they are named)
  pharmacy = 11,
  // again these could just be a random building with nothing other than a sign like "early years" or something, they better have a name or it will be ambiguous
  kindergarten = 12,
  // same thing, often just a building
  bar = 13,
  // this could be very very big like a college campus, i think we should remove this
  hospital = 14,
  // same as bar could be just a building need at least a name to have a chance of seeing it
  pub = 16,
  // generic, could be manyy many types of clinics also often there are multiple of these located in same "complex", you might have a dentist, orthodontist, ear nose throat and eye specialists all next to each other. 
  clinic = 17,
  // on the fence about this one, to me this would be in the same category as "museum" or something where its so specialized maybe it doesnt need a name. what keeps me from say ing that is you often cant tell from the outside what function these amenities serve until you read the name
  theatre = 16,
  // this could work without a name in my opinion, they often have a board out front that shows what movies are playing which would probably remove the need for a name
  cinema = 17,
  // these are usually quite large as well but not usually multiple buildings, how often are they located near one another though? if its often perhaps we need a name to differentiate between this casino and the next one down on the next block? maybe that argument is a good argument for everything has to have a name?
  casino = 18,
};
```

vesperlou replied:
I've added the remarks. I feel this is going to be a long term discussion and decision :)

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -1,10 +1,61 @@
+
+#include "mjolnir/osmpbfparser.h"
 #include "mjolnir/util.h"
+#include <boost/property_tree/ptree.hpp>
```
you already have this in the header
```suggestion
```

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -1,10 +1,61 @@
+  node_callback(const uint64_t /*osmid*/, double lng, double lat, const OSMPBF::Tags& tags) override {
+    auto iter = tags.find("amenity");
+    if (iter != tags.cend() && !iter->second.empty()) {
+      valhalla::mjolnir::LandmarkType type = static_cast<valhalla::mjolnir::LandmarkType>(0);
+      try {
+        type = valhalla::mjolnir::string_to_landmark_type(iter->second);
+      } catch (const std::runtime_error& e) { return; }
+
+      std::string name = "";
+      auto it = tags.find("name");
+      if (it != tags.cend() && !it->second.empty()) {
+        name = it->second;
+      }
+
+      // insert parsed landmark directly into database
+      db_.insert_landmark(name, type, lng, lat);
```
you can avoid this clunkiness by doing the whole thing inside the try catch. also dont use the name `type` as a variable name its just too close to keywords and likely other things it could collide with.

```suggestion
      try {
        auto landmark_type = valhalla::mjolnir::string_to_landmark_type(iter->second);
      
        std::string name = "";
        auto it = tags.find("name");
        if (it != tags.cend() && !it->second.empty()) {
          name = it->second;
        }

        // insert parsed landmark directly into database
        db_.insert_landmark(name, landmark_type, lng, lat);
      } catch (const std::runtime_error& e) { return; }
```

vesperlou replied:
nice suggestion thanks

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -1,10 +1,61 @@
+    }
+  }
+
+  virtual void changeset_callback(const uint64_t changeset_id) override {
```
```suggestion
  virtual void changeset_callback(const uint64_t /*changeset_id*/) override {
```

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -168,5 +221,44 @@ std::vector<Landmark> LandmarkDatabase::get_landmarks_in_bounding_box(const doub
 
+bool BuildLandmarkFromPBF(const boost::property_tree::ptree& pt,
+                          const std::vector<std::string>& input_files) {
+  // parse config to get landmark database
+  auto db = pt.get_optional<std::string>("landmarks");
+  if (!db) {
+    LOG_ERROR("Landmarks config info not found. Landmarks builder will not be created.");
+    return false;
+  }
```
its not bad to have this but its not needed exactly.. the config generator will always put a value here, one option would be to do:

```suggestion
  auto db = pt.get<std::string>("landmarks", "");
```

then when we got to make the db it should fail with an invalid path

vesperlou replied:
much more concise

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -168,5 +221,44 @@ std::vector<Landmark> LandmarkDatabase::get_landmarks_in_bounding_box(const doub
+    return false;
+  }
+
+  const filesystem::path parent_dir = filesystem::path(*db).parent_path();
+  if (!filesystem::exists(parent_dir) && !filesystem::create_directories(parent_dir)) {
+    LOG_ERROR("Can't create parent directory " + parent_dir.string());
+    return false;
+  }
```
ok so we have the removal of existing db in the constructor then we need this create the directory if it doesnt exist also in the constructor, it doesnt make sense for the two things to be decoupled

vesperlou replied:
make sense. done.

kevinkreiser commented on test/gurka/test_landmarks.cc:
```diff
@@ -1,61 +1,128 @@
-  static void SetUpTestSuite() {
-    LandmarkDatabase db(db_name, false);
+  const gurka::nodes nodes = {
+      {"A", {{"name", ""}, {"amenity", "bar"}}},
```
in real life osm will not store a tag without a value, so if you want to test this then we should just omit the name tag completely:

```suggestion
      {"A", {{"amenity", "bar"}}},
```

vesperlou replied:
good to know!

vesperlou commented on test/gurka/test_landmarks.cc:
```diff
@@ -1,61 +1,128 @@
+      {"D", {{"name", "wan da"}, {"amenity", "cinema"}}},
+      {"E", {{"name", "zhong lou"}, {"amenity", "monument"}}}, // not in list, shouldn't be stored
+      // non-landmark nodes
+      {"a", {{"name", "gong ce"}, {"amenity", ""}}},
+      {"b", {{"name", "la ji tong"}}},
+      {"c", {{"name", "hua yuan"}}},
+      {"d", {{"name", ""}, {"amenity", ""}}},
```
sure it makes sense to have extra tags and thanks for adding those

kevinkreiser commented on valhalla/mjolnir/landmark_builder.h:
```diff
@@ -0,0 +1,116 @@
+  hospital = 14,
+  // same as bar could be just a building need at least a name to have a chance of seeing it
+  pub = 16,
+  // generic, could be manyy many types of clinics also often there are multiple of these located in
```
```suggestion
  // generic, could be many many types of clinics also often there are multiple of these located in
```

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -16,12 +66,20 @@ struct LandmarkDatabase::db_pimpl {
   db_pimpl(const std::string& db_name, bool read_only) : insert_stmt(nullptr) {
-    // figure out if we need to create it or can just open it up
-    auto flags = read_only ? SQLITE_OPEN_READONLY : SQLITE_OPEN_READWRITE;
+    // create parent directory if not exist
```
```suggestion
    // create parent directory if it doesn't exist
```

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -16,12 +66,20 @@ struct LandmarkDatabase::db_pimpl {
         throw std::logic_error("Cannot open sqlite database in read-only mode if it does not exist");
-      flags |= SQLITE_OPEN_CREATE;
+    } else if (!read_only) {
+      std::filesystem::remove(db_name);
```
long story but we only just went to c++17 which as non-experimental support for filesystem. before that we use valhalla::filesystem, our internal home grown version. we've not migrated to std::filesystem yet so lets use our own until we do so

```suggestion
      filesystem::remove(db_name);
```

kevinkreiser commented on src/mjolnir/landmark_builder.cc:
```diff
@@ -1,10 +1,60 @@
+
+#include "mjolnir/osmpbfparser.h"
 #include "mjolnir/util.h"
+#include <filesystem>
```
```suggestion
```

kevinkreiser reviewed (changes requested):
i had some tiny nit picks yet and then this is ready to go

kevinkreiser commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@
    * ADDED: `source_percent_along` & `target_percent_along` to /trace_attributes JSON response [#4199](https://github.com/valhalla/valhalla/pull/4199)
    * ADDED: sqlite database to store landmarks along with interfaces of insert and bounding box queries [#4189](https://github.com/valhalla/valhalla/pull/4189)
    * CHANGED: refactor landmark database interface to use a pimpl [#4202](https://github.com/valhalla/valhalla/pull/4202)
+   * ADDED: add landmark parser to parse POI from pbf files and store them as landmarks in landmark database [#4201](https://github.com/valhalla/valhalla/pull/4201)
```
```suggestion
   * ADDED: add `valhalla_build_landmarks` to parse POIs from osm pbfs and store them as landmarks in the landmark sqlite database [#4201](https://github.com/valhalla/valhalla/pull/4201)
```
